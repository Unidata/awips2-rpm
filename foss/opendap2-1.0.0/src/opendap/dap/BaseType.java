/////////////////////////////////////////////////////////////////////////////
// This file is part of the "Java-DAP" project, a Java implementation
// of the OPeNDAP Data Access Protocol.
//
// Copyright (c) 2010, OPeNDAP, Inc.
// Copyright (c) 2002,2003 OPeNDAP, Inc.
// 
// Author: James Gallagher <jgallagher@opendap.org>
// 
// All rights reserved.
// 
// Redistribution and use in source and binary forms,
// with or without modification, are permitted provided
// that the following conditions are met:
// 
// - Redistributions of source code must retain the above copyright
//   notice, this list of conditions and the following disclaimer.
// 
// - Redistributions in binary form must reproduce the above copyright
//   notice, this list of conditions and the following disclaimer in the
//   documentation and/or other materials provided with the distribution.
// 
// - Neither the name of the OPeNDAP nor the names of its contributors may
//   be used to endorse or promote products derived from this software
//   without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
// TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
// PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
/////////////////////////////////////////////////////////////////////////////


package opendap.dap;

import opendap.util.EscapeStrings;

import java.io.*;
import java.util.Enumeration;

/**
 * This abstract class defines the basic data type features for the OPeNDAP data
 * access protocol (DAP) data types. All of the DAP type classes
 * (<code>DFloat64</code>, <code>DArray</code>, etc.) subclass it or one of
 * its two abstract descendents, <code>DVector</code> or
 * <code>DConstructor</code>.
 * <p/>
 * These classes and their methods give a user the capacity to set up
 * sophisticated data types. They do <em>not</em> provide sophisticated ways to
 * access and use this data. On the server side, in many cases, the class
 * instances will have no data in them at all until the <code>serialize</code>
 * method is called to send data to the client. On the client side, most OPeNDAP
 * application programs will unpack the data promptly into whatever local
 * data structure the programmer deems the most useful.
 * <p/>
 * Descendents of this class should implement the <code>ClientIO</code>
 * interface.  That interface defines a <code>deserialize</code> method used
 * by a OPeNDAP client to retrieve the variable's declaration and value(s) from
 * a OPeNDAP server.
 *
 * @author jehamby
 * @version $Revision: 25753 $
 * @see DDS
 * @see ClientIO
 */
public abstract class BaseType implements Cloneable, Serializable {

    static final long serialVersionUID = 1;

    /**
     * The name of this variable.
     */
    private String _name;

    /**
     * The parent (container class) of this object, if one exists
     */
    private BaseType _myParent;

    /**
     * The Attribute Table used to contain attributes specific to this
     * instance of a BaseType variable. This is the repository for
     * "Semantic Metadata"
     */
    private Attribute _attr;
    private AttributeTable _attrTbl;

    /**
     * Constructs a new <code>BaseType</code> with no name.
     */
    public BaseType() {
        this(null);
    }

    /**
     * Constructs a new <code>BaseType</code> with name <code>n</code>.
     *
     * @param n the name of the variable.
     */
    public BaseType(String n) {
        this(n, true);

    }

    /**
     * Constructs a new <code>BaseType</code> with name <code>n</code>.
     *
     * @param n the name of the variable.
     */
    public BaseType(String n, boolean decodeName) {

        if (decodeName)
            _name = EscapeStrings.www2id(n);
        else
            _name = n;


        _myParent = null;
        _attrTbl = new AttributeTable(_name);
        _attr = new Attribute(_name, _attrTbl);
    }

    /**
     * Returns a clone of this <code>BaseType</code>.  A deep copy is performed
     * on all data inside the variable.
     *
     * @return a clone of this <code>BaseType</code>.
     */
    public Object clone() {
        try {
            BaseType bt = (BaseType) super.clone();
            bt._name = _name;
            bt._attrTbl = (AttributeTable) this._attrTbl.clone();
            bt._attr = new Attribute(getName(), bt._attrTbl);
            return bt;
        } catch (CloneNotSupportedException e) {
            // this shouldn't happen, since we are Cloneable
            throw new InternalError();
        }
    }


    /**
     * Returns the unencoded name of the class instance.
     *
     * @return the name of the class instance.
     */
    public final String getClearName() {
        return _name;
    }


    /**
     * Returns the WWW encoded name of the class instance.
     *
     * @return the name of the class instance.
     */
    public final String getName() {
        return EscapeStrings.id2www(_name);
    }

    /**
     * Sets the name of the class instance.
     *
     * @param n the name of the class instance.
     */
    public final void setName(String n) {

        setClearName(EscapeStrings.www2id(n));

    }

    /**
     * Sets the unencoded name of the class instance.
     *
     * @param n the unencoded name of the class instance.
     */
    public final void setClearName(String n) {
        _name = n;
        _attr.setClearName(n);
        _attrTbl.setClearName(n);
    }

    /**
     * Returns the OPeNDAP type name of the class instance as a <code>String</code>.
     *
     * @return the OPeNDAP type name of the class instance as a <code>String</code>.
     */
    abstract public String getTypeName();

    /**
     * Returns the number of variables contained in this object. For simple and
     * vector type variables, it always returns 1. To count the number
     * of simple-type variable in the variable tree rooted at this variable, set
     * <code>leaves</code> to <code>true</code>.
     *
     * @param leaves If true, count all the simple types in the `tree' of
     *               variables rooted at this variable.
     * @return the number of contained variables.
     */
    public int elementCount(boolean leaves) {
        return 1;
    }

    /**
     * Returns the number of variables contained in this object. For simple and
     * vector type variables, it always returns 1.
     *
     * @return the number of contained variables.
     */
    public final int elementCount() {
        return elementCount(false);
    }


    /**
     * Write the variable's declaration in a C-style syntax. This
     * function is used to create textual representation of the Data
     * Descriptor Structure (DDS).  See <em>The OPeNDAP User Manual</em> for
     * information about this structure.
     *
     * @param os          The <code>PrintWriter</code> on which to print the
     *                    declaration.
     * @param space       Each line of the declaration will begin with the
     *                    characters in this string.  Usually used for leading spaces.
     * @param print_semi  a boolean value indicating whether to print a
     *                    semicolon at the end of the declaration.
     * @param constrained a boolean value indicating whether to print
     *                    the declartion dependent on the projection information. <b>This
     *                    is only used by Server side code.</b>
     * @see DDS
     */
    public void printDecl(PrintWriter os, String space,
                          boolean print_semi, boolean constrained) {

        //System.out.println("BaseType.printDecl()...");
        os.print(space + getTypeName() + " " + getName());
        if (print_semi)
            os.println(";");
    }

    /**
     * Write the variable's declaration in a C-style syntax. This
     * function is used to create textual representation of the Data
     * Descriptor Structure (DDS).  See <em>The OPeNDAP User Manual</em> for
     * information about this structure.
     *
     * @param os         The <code>PrintWriter</code> on which to print the
     *                   declaration.
     * @param space      Each line of the declaration will begin with the
     *                   characters in this string.  Usually used for leading spaces.
     * @param print_semi a boolean value indicating whether to print a
     *                   semicolon at the end of the declaration.
     * @see DDS
     */
    public void printDecl(PrintWriter os, String space,
                          boolean print_semi) {

        printDecl(os, space, print_semi, false);

    }

    /**
     * Print the variable's declaration.  Same as
     * <code>printDecl(os, space, true)</code>.
     *
     * @param os    The <code>PrintWriter</code> on which to print the
     *              declaration.
     * @param space Each line of the declaration will begin with the
     *              characters in this string.  Usually used for leading spaces.
     * @see DDS#print(PrintWriter)
     */
    public final void printDecl(PrintWriter os, String space) {
        printDecl(os, space, true, false);
    }

    /**
     * Print the variable's declaration.  Same as
     * <code>printDecl(os, "    ", true)</code>.
     *
     * @param os The <code>PrintWriter</code> on which to print the
     *           declaration.
     * @see DDS#print(PrintWriter)
     */
    public final void printDecl(PrintWriter os) {
        printDecl(os, "    ", true, false);
    }

    /**
     * Print the variable's declaration using <code>OutputStream</code>.
     *
     * @param os         The <code>OutputStream</code> on which to print the
     *                   declaration.
     * @param space      Each line of the declaration will begin with the
     *                   characters in this string.  Usually used for leading spaces.
     * @param print_semi a boolean value indicating whether to print a
     *                   semicolon at the end of the declaration.
     * @see DDS#print(PrintWriter)
     */
    public final void printDecl(OutputStream os, String space,
                                boolean print_semi, boolean constrained) {
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(os)));
        printDecl(pw, space, print_semi, constrained);
        pw.flush();
    }

    /**
     * Print the variable's declaration using <code>OutputStream</code>.
     *
     * @param os         The <code>OutputStream</code> on which to print the
     *                   declaration.
     * @param space      Each line of the declaration will begin with the
     *                   characters in this string.  Usually used for leading spaces.
     * @param print_semi a boolean value indicating whether to print a
     *                   semicolon at the end of the declaration.
     * @see DDS#print(PrintWriter)
     */
    public final void printDecl(OutputStream os, String space,
                                boolean print_semi) {
        printDecl(os, space, print_semi, false);
    }

    /**
     * Print the variable's declaration.  Same as
     * <code>printDecl(os, space, true)</code>.
     *
     * @param os    The <code>OutputStream</code> on which to print the
     *              declaration.
     * @param space Each line of the declaration will begin with the
     *              characters in this string.  Usually used for leading spaces.
     * @see DDS#print(PrintWriter)
     */
    public final void printDecl(OutputStream os, String space) {
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(os)));
        printDecl(pw, space);
        pw.flush();
    }

    /**
     * Print the variable's declaration.  Same as
     * <code>printDecl(os, "    ", true)</code>.
     *
     * @param os The <code>OutputStream</code> on which to print the
     *           declaration.
     * @see DDS#print(PrintWriter)
     */
    public final void printDecl(OutputStream os) {
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(os)));
        printDecl(pw);
        pw.flush();
    }


    /**
     * Prints the value of the variable, with its declaration.  This
     * function is primarily intended for debugging OPeNDAP applications and
     * text-based clients such as geturl.
     *
     * @param os           the <code>PrintWriter</code> on which to print the value.
     * @param space        this value is passed to the <code>printDecl</code> method,
     *                     and controls the leading spaces of the output.
     * @param print_decl_p a boolean value controlling whether the
     *                     variable declaration is printed as well as the value.
     */
    abstract public void printVal(PrintWriter os, String space,
                                  boolean print_decl_p);

    /**
     * Print the variable's value.  Same as
     * <code>printVal(os, space, true)</code>.
     *
     * @param os    the <code>PrintWriter</code> on which to print the value.
     * @param space this value is passed to the <code>printDecl</code> method,
     *              and controls the leading spaces of the output.
     * @see DataDDS#printVal(PrintWriter)
     */
    public final void printVal(PrintWriter os, String space) {
        printVal(os, space, true);
    }

    /**
     * Print the variable's value using <code>OutputStream</code>.
     *
     * @param os           the <code>OutputStream</code> on which to print the value.
     * @param space        this value is passed to the <code>printDecl</code> method,
     *                     and controls the leading spaces of the output.
     * @param print_decl_p a boolean value controlling whether the
     *                     variable declaration is printed as well as the value.
     * @see DataDDS#printVal(PrintWriter)
     */
    public final void printVal(OutputStream os, String space,
                               boolean print_decl_p) {
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(os)));
        printVal(pw, space, print_decl_p);
        pw.flush();
    }

    /**
     * Print the variable's value using <code>OutputStream</code>.
     *
     * @param os    the <code>OutputStream</code> on which to print the value.
     * @param space this value is passed to the <code>printDecl</code> method,
     *              and controls the leading spaces of the output.
     * @see DataDDS#printVal(PrintWriter)
     */
    public final void printVal(OutputStream os, String space) {
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(os)));
        printVal(pw, space);
        pw.flush();
    }

    /**
     * Checks for internal consistency.  This is important to check for complex
     * constructor classes.
     * For example, an <code>DInt32</code> instance would return false if it had
     * no name defined.  A <code>DGrid</code> instance might return false for
     * more complex reasons, such as having Map arrays of the wrong
     * size or shape.
     * <p/>
     * This method is used by the <code>DDS</code> class, and will rarely, if
     * ever, be explicitly called by a OPeNDAP application program.  A
     * variable must pass this test before it is sent, but there may be
     * many other stages in a retrieve operation where it would fail.
     *
     * @param all For complex constructor types (
     *            <code>DGrid</code>, <code>DSequence</code>, <code>DStructure</code>),
     *            this flag indicates whether to check the
     *            semantics of the member variables, too.
     * @throws BadSemanticsException if semantics are bad, explains why.
     * @see DDS#checkSemantics(boolean)
     */
    public void checkSemantics(boolean all)
            throws BadSemanticsException {
        if (_name == null)
            throw new BadSemanticsException("BaseType.checkSemantics(): Every variable must have a name");
    }

    /**
     * Check semantics.  Same as <code>checkSemantics(false)</code>.
     *
     * @throws BadSemanticsException if semantics are bad, explains why.
     * @see BaseType#checkSemantics(boolean)
     */
    public final void checkSemantics() throws BadSemanticsException {
        checkSemantics(false);
    }

    /**
     * Constructs a new <code>PrimitiveVector</code> object optimized for the
     * particular data type of this <code>BaseType</code>.  For example, a
     * <code>DByte</code> class would create a new
     * <code>BytePrimitiveVector</code> in this call.  This allows for a very
     * optimized, yet type-safe, implementation of <code>DVector</code>
     * functionality.  For non-primitive types, such as
     * <code>DArray</code>, <code>DGrid</code>, <code>DSequence</code>, and
     * <code>DStructure</code>, the default implementation returns a
     * <code>BaseTypePrimitiveVector</code> object which can
     * deserialize an array of complex types.
     *
     * @return a new <code>PrimitiveVector</code> object for the variable type.
     */
    public PrimitiveVector newPrimitiveVector() {
        return new BaseTypePrimitiveVector(this);
    }


    public void setParent(BaseType bt) {
        _myParent = bt;
    }

    public BaseType getParent() {
        return (_myParent);
    }

    public String getLongName() {

        boolean done = false;

        BaseType parent = _myParent;

        String longName = _name;


        while (parent != null) {
            longName = parent.getName() + "." + longName;
            parent = parent.getParent();
        }
        return (longName);
    }

    // **************************************************************
    //
    // Attribute Table Methods
    //
    // ..............................................................


    public boolean hasAttributes() {


        Enumeration e = _attrTbl.getNames();

        if (e.hasMoreElements())
            return (true);

        return (false);


    }


    public Attribute getAttribute() {
        return _attr;
    }

    public AttributeTable getAttributeTable() {
        return _attrTbl;
    }

    public void addAttributeAlias(String alias, String attributeName)
            throws DASException {
        _attrTbl.addAlias(alias, attributeName);
    }

    public void appendAttribute(String name, int type, String value, boolean check)
            throws DASException {
        _attrTbl.appendAttribute(name, type, value, check);
    }

    public void appendAttribute(String name, int type, String value)
            throws DASException {
        _attrTbl.appendAttribute(name, type, value);
    }

    public void addAttributeContainer(AttributeTable at)
            throws AttributeExistsException {
        _attrTbl.addContainer(at.getName(), at);
    }

    public AttributeTable appendAttributeContainer(String name) {
        return (_attrTbl.appendContainer(name));
    }

    public void delAttribute(String name) {
        _attrTbl.delAttribute(name);
    }

    public void delAttribute(String name, int i) throws DASException {
        _attrTbl.delAttribute(name, i);
    }

    public Attribute getAttribute(String name) {
        return (_attrTbl.getAttribute(name));
    }

    public Enumeration getAttributeNames() {
        return (_attrTbl.getNames());
    }

    public void printAttributes(OutputStream os) {
        _attrTbl.print(os);
    }

    public void printAttributes(OutputStream os, String pad) {
        _attrTbl.print(os, pad);
    }

    public void printAttributes(PrintWriter pw) {
        _attrTbl.print(pw);
    }

    public void printAttributes(PrintWriter pw, String pad) {
        _attrTbl.print(pw, pad);
    }


    /**
     *
     * @param os
     * @opendap.ddx.experimental
     */
    public void printXML(OutputStream os) {
        printXML(os, "");
    }


    /**
     *
     * @param os
     * @param pad
     * @opendap.ddx.experimental
     */
    public void printXML(OutputStream os, String pad) {
        PrintWriter pw = new PrintWriter(new BufferedWriter(new OutputStreamWriter(os)));
        printXML(pw, pad, false);
        pw.flush();
    }


    /**
     *
     * @param pw
     * @opendap.ddx.experimental
     */
    public void printXML(PrintWriter pw) {
        printXML(pw, "");
    }


    /**
     *
     * @param pw
     * @param pad
     * @opendap.ddx.experimental
     */
    public void printXML(PrintWriter pw, String pad) {
        printXML(pw, pad, false);

    }

    /**
     *
     * @param pw
     * @param pad
     * @param constrained
     * @opendap.ddx.experimental
     */
    public void printXML(PrintWriter pw, String pad, boolean constrained) {

        pw.print(pad + "<" + getTypeName());
        if (_name != null) {
            pw.print(" name=\"" +
                    opendap.dap.XMLparser.DDSXMLParser.normalizeToXML(_name) + "\"");
        }

        Enumeration e = getAttributeNames();
        if (e.hasMoreElements()) {
            pw.println(">");
            while (e.hasMoreElements()) {
                String aName = (String) e.nextElement();

                Attribute a = getAttribute(aName);
                if(a!=null)
                    a.printXML(pw, pad + "\t", constrained);

            }
            pw.println(pad + "</" + getTypeName() + ">");
        } else {
            pw.println("/>");
        }


    }


}


