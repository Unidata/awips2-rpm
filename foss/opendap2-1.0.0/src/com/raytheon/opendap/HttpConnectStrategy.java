/**
 * This software was developed and / or modified by Raytheon Company,
 * pursuant to Contract DG133W-05-CQ-1067 with the US Government.
 *
 * U.S. EXPORT CONTROLLED TECHNICAL DATA
 * This software product contains export-restricted data whose
 * export/transfer/disclosure is restricted by U.S. law. Dissemination
 * to non-U.S. persons whether in the United States or abroad requires
 * an export license or other authorization.
 *
 * Contractor Name:        Raytheon Company
 * Contractor Address:     6825 Pine Street, Suite 340
 *                         Mail Stop B8
 *                         Omaha, NE 68106
 *                         402.291.0100
 *
 * See the AWIPS II Master Rights File ("Master Rights File.pdf") for
 * further licensing information.
 **/
package com.raytheon.opendap;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;

import opendap.dap.DAP2Exception;
import opendap.dap.ServerVersion;

/**
 * A strategy for making HTTP requests.
 *
 * <pre>
 *
 * SOFTWARE HISTORY
 *
 * Date          Ticket#  Engineer  Description
 * ------------- -------- --------- -----------------------------
 * Jul 06, 2012  634      djohnson  Initial creation
 * Jun 12, 2018  7320     rjpeter   Removed unavailable options.
 *
 * </pre>
 *
 * @author djohnson
 */
public interface HttpConnectStrategy {

    String DODS_ERROR = "dods_error";

    /**
     * Get the {@link InputStream} for the {@link URL}.
     *
     * @param url
     *            the url
     * @return the input stream
     * @throws IOException
     *             on error
     * @throws DODSException
     */
    InputStream getInputStream(java.net.URL url)
            throws IOException, DAP2Exception;

    /**
     * Return the server version.
     *
     * @return the server version
     */
    ServerVersion getServerVersion();
}
