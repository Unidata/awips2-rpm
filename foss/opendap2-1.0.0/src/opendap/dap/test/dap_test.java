/////////////////////////////////////////////////////////////////////////////
// This file is part of the "Java-DAP" project, a Java implementation
// of the OPeNDAP Data Access Protocol.
//
// Copyright (c) 2010, OPeNDAP, Inc.
// Copyright (c) 2002,2003 OPeNDAP, Inc.
// 
// Author: James Gallagher <jgallagher@opendap.org>
// 
// All rights reserved.
// 
// Redistribution and use in source and binary forms,
// with or without modification, are permitted provided
// that the following conditions are met:
// 
// - Redistributions of source code must retain the above copyright
//   notice, this list of conditions and the following disclaimer.
// 
// - Redistributions in binary form must reproduce the above copyright
//   notice, this list of conditions and the following disclaimer in the
//   documentation and/or other materials provided with the distribution.
// 
// - Neither the name of the OPeNDAP nor the names of its contributors may
//   be used to endorse or promote products derived from this software
//   without specific prior written permission.
// 
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
// TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
// PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
/////////////////////////////////////////////////////////////////////////////

package opendap.dap.test;

import opendap.dap.*;

import java.util.Vector;

public class dap_test {
    private static final String prompt = "dds-test: ";
    private static final String version = "version 0.1";

    public static void main(String args[]) {
        test_das();
        test_dds();
    }

    private static void test_das() {
        System.out.println("DAS test:");

        DAS table = new DAS();
        AttributeTable at = new AttributeTable("DAS Test");
        try {
            at.appendAttribute("Authors", Attribute.STRING, "jehamby");
            at.appendAttribute("Authors", Attribute.STRING, "jimg");
            at.appendAttribute("Authors", Attribute.STRING, "ndp");
            at.addAlias("Creators", "Authors");
            AttributeTable cont = at.appendContainer("Container");
            cont.appendAttribute("Numbers", Attribute.INT32, "123");
            cont.appendAttribute("Floats", Attribute.FLOAT64, "456.0");
            table.addAttributeTable("table1", at);
            at = new AttributeTable("Nested Attribute Collection");
            at.appendAttribute("another", Attribute.BYTE, "12");
            table.addAttributeTable("table2", at);

            table.resolveAliases();
        }
        catch (DASException e) {
            System.out.println("Error constructing DAS: " + e);
        }

        try {
            // print the table
            table.print(System.out);

            // get a name from the table
            String author2 = table.getAttributeTable("table1").getAttribute("Creators").getValueAt(1);
            System.out.println("author2 = " + author2);

            // test Cloneable interface
            DAS table2 = (DAS) table.clone();
            System.out.println("\nCloned table:");
            table2.print(System.out);

            // add an attribute to table1 and verify it's not in table 2
            at = new AttributeTable("DAS clone test");
            table.addAttributeTable("empty", at);
            if (table2.getAttributeTable("empty") != null)
                System.out.println("DAS clone failed");
            else
                System.out.println("DAS clone passed");
        }
        catch (DASException e) {
            System.out.println("Error Using DAS: " + e);
        }
    }

    private static void test_dds() {
        System.out.println("DDS test:");

        DataDDS table = new DataDDS(new ServerVersion(2, 16));
        table.setName("test_table");

        // add variables to it
        DUInt32 myUInt = new DUInt32("myUInt");
        myUInt.setValue(42);
        table.addVariable(myUInt);

        // note that arrays and lists take their name from the addVariable method
        DArray myArray = new DArray();
        myArray.addVariable(new DByte("myArray"));
        myArray.appendDim(10, "dummy");
        myArray.setLength(10);
        BytePrimitiveVector bpv = (BytePrimitiveVector) myArray.getPrimitiveVector();
        for (int i = 0; i < 10; i++)
            bpv.setValue(i, (byte) (i * 10));
        table.addVariable(myArray);

        DStructure myStructure = new DStructure("myStructure");
        DFloat64 structFloat = new DFloat64("structFloat");
        structFloat.setValue(42.0);
        myStructure.addVariable(structFloat);
        DString structString = new DString("structString");
        structString.setValue("test value");
        myStructure.addVariable(structString);
        table.addVariable(myStructure);

        DGrid myGrid = new DGrid("myGrid");
        DArray gridArray = (DArray) myArray.clone();
        gridArray.setName("gridArray");
        myGrid.addVariable(gridArray, DGrid.ARRAY);
        DArray gridMap = new DArray();
        gridMap.addVariable(new DInt32("gridMap"));
        gridMap.appendDim(10, "dummy");
        gridMap.setLength(10);
        Int32PrimitiveVector ipv = (Int32PrimitiveVector) gridMap.getPrimitiveVector();
        for (int i = 0; i < 10; i++)
            ipv.setValue(i, i * 10);
        myGrid.addVariable(gridMap, DGrid.MAPS);
        table.addVariable(myGrid);

        // this is the one case where two OPeNDAP variables can have the same name:
        // each row should have the same exact variables (name doesn't matter)
        DSequence mySequence = new DSequence("mySequence");
        mySequence.addVariable(new DInt32("seqInt32"));
        mySequence.addVariable(new DString("seqString"));
        Vector<BaseType> seqRow = new Vector<BaseType>();  // a row of the sequence
        DInt32 seqVar1 = new DInt32("seqInt32");
        seqVar1.setValue(1);
        seqRow.addElement(seqVar1);
        DString seqVar2 = new DString("seqString");
        seqVar2.setValue("string");
        seqRow.addElement(seqVar2);
        mySequence.addRow(seqRow);
        seqRow = new Vector<BaseType>();  // add a second row
        seqVar1 = new DInt32("seqInt32");
        seqVar1.setValue(3);
        seqRow.addElement(seqVar1);
        seqVar2 = new DString("seqString");
        seqVar2.setValue("another string");
        seqRow.addElement(seqVar2);
        mySequence.addRow(seqRow);
        table.addVariable(mySequence);

        try {
            table.checkSemantics();
            System.out.println("DDS passed semantic check");
        }
        catch (BadSemanticsException e) {
            System.out.println("DDS failed semantic check:\n" + e);
        }

        try {
            table.checkSemantics(true);
            System.out.println("DDS passed full semantic check");
        }
        catch (BadSemanticsException e) {
            System.out.println("DDS failed full semantic check:\n" + e);
        }

        // print the declarations
        System.out.println("declarations:");
        table.print(System.out);

        // print the data
        System.out.println("\nData:");
        table.printVal(System.out);
        System.out.println();

        // and read it programmatically
        try {
            int testValue1 = ((DUInt32) table.getVariable("myUInt")).getValue();
            System.out.println("myUInt = " + testValue1);
            byte testValue2 = ((BytePrimitiveVector) ((DArray) table.getVariable("myArray")).getPrimitiveVector()).getValue(5);
            System.out.println("myArray[5] = " + testValue2);
            double testValue4 = ((DFloat64) ((DStructure) table.getVariable("myStructure")).getVariable("structFloat")).getValue();
            System.out.println("myStructure.structFloat = " + testValue4);
            int testValue5 = ((Int32PrimitiveVector) ((DArray) ((DGrid) table.getVariable("myGrid")).getVariable("gridMap")).getPrimitiveVector()).getValue(5);
            System.out.println("myGrid.gridMap[5] = " + testValue5);
            String testValue7 = ((DString) ((DSequence) table.getVariable("mySequence")).getVariable(0, "seqString")).getValue();
            System.out.println("mySequence[0].seqString = " + testValue7);
        }
        catch (NoSuchVariableException e) {
            System.out.println("Error getting variable:\n" + e);
        }
        System.out.println();

        DataDDS table2 = (DataDDS) table.clone();  // test Cloneable interface
        try {
            table2.checkSemantics();
            System.out.println("DDS passed semantic check");
        }
        catch (BadSemanticsException e) {
            System.out.println("DDS failed semantic check:\n" + e);
        }

        try {
            table2.checkSemantics(true);
            System.out.println("DDS passed full semantic check");
        }
        catch (BadSemanticsException e) {
            System.out.println("DDS failed full semantic check:\n" + e);
        }

        // print the declarations and data
        System.out.println("clone declarations:");
        table2.print(System.out);
        System.out.println("\nData:");
        table2.printVal(System.out);
        System.out.println();

        // add some values to the original table
        DInt32 myNewInt = new DInt32("myNewInt");
        myNewInt.setValue(420);
        table.addVariable(myNewInt);

        // verify that they aren't in the cloned table
        try {
            DInt32 testInt = (DInt32) table2.getVariable("myNewInt");
            System.out.println("Error: value from table in table2  "+testInt);
        }
        catch (NoSuchVariableException e) {
            System.out.println("Variable cloning looks good");
        }
    }
}


