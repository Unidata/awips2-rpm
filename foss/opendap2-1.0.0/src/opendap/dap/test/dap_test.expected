DAS test:
Attributes {
    table1 {
        String Authors jehamby, jimg;
        Alias Creators Authors;
        Container {
            Int32 Numbers 123;
            Float64 Floats 456.0;
        }
    }
    table2 {
        Byte another 12;
    }
}
author2 = jimg

Cloned table:
Attributes {
    table1 {
        String Authors jehamby, jimg;
        Alias Creators Authors;
        Container {
            Int32 Numbers 123;
            Float64 Floats 456.0;
        }
    }
    table2 {
        Byte another 12;
    }
}
DAS clone passed
DDS test:
DDS passed semantic check
DDS passed full semantic check
declarations:
Dataset {
    UInt32 myUInt;
    Byte myArray[dummy = 10];
    List Url myList;
    Structure {
        Float64 structFloat;
        String structString;
    } myStructure;
    Grid {
     ARRAY:
        Byte myArray[dummy = 10];
     MAPS:
        Int32 gridMap[dummy = 10];
    } myGrid;
    Sequence {
        Int32 seqInt32;
        String seqString;
    } mySequence;
} test_table;

Data:
UInt32 myUInt = 42;
Byte myArray[dummy = 10] = {0, 10, 20, 30, 40, 50, 60, 70, 80, 90};
List Url myList = { "http://0", "http://1", "http://2", "http://3", "http://4", "http://5", "http://6", "http://7", "http://8", "http://9"};
Structure {
    Float64 structFloat;
    String structString;
} myStructure = { 42.0, "test value" };
Grid {
 ARRAY:
    Byte myArray[dummy = 10];
 MAPS:
    Int32 gridMap[dummy = 10];
} myGrid = { ARRAY: {0, 10, 20, 30, 40, 50, 60, 70, 80, 90} MAPS: {0, 10, 20, 30, 40, 50, 60, 70, 80, 90} };
Sequence {
    Int32 seqInt32;
    String seqString;
} mySequence = { { 1, "string" }, { 3, "another string" } };


myUInt = 42
myArray[5] = 50
myList[5] = http://5
myStructure.structFloat = 42.0
myGrid.gridMap[5] = 50
mySequence[0].seqString = string

DDS passed semantic check
DDS passed full semantic check
clone declarations:
Dataset {
    UInt32 myUInt;
    Byte myArray[dummy = 10];
    List Url myList;
    Structure {
        Float64 structFloat;
        String structString;
    } myStructure;
    Grid {
     ARRAY:
        Byte myArray[dummy = 10];
     MAPS:
        Int32 gridMap[dummy = 10];
    } myGrid;
    Sequence {
        Int32 seqInt32;
        String seqString;
    } mySequence;
} test_table;

Data:
UInt32 myUInt = 42;
Byte myArray[dummy = 10] = {0, 10, 20, 30, 40, 50, 60, 70, 80, 90};
List Url myList = { "http://0", "http://1", "http://2", "http://3", "http://4", "http://5", "http://6", "http://7", "http://8", "http://9"};
Structure {
    Float64 structFloat;
    String structString;
} myStructure = { 42.0, "test value" };
Grid {
 ARRAY:
    Byte myArray[dummy = 10];
 MAPS:
    Int32 gridMap[dummy = 10];
} myGrid = { ARRAY: {0, 10, 20, 30, 40, 50, 60, 70, 80, 90} MAPS: {0, 10, 20, 30, 40, 50, 60, 70, 80, 90} };
Sequence {
    Int32 seqInt32;
    String seqString;
} mySequence = { { 1, "string" }, { 3, "another string" } };


Variable cloning looks good
